ElementTypeValueList

tlorantfy opened the pull request:
A ValueList component to choose an Archicad element type from dropdown list.

https://github.com/user-attachments/assets/17343b53-6967-48ef-a78f-07089a759f0d

thekaushikls commented:
@tlorantfy, do you think it is better to have _"the feature"_ inside the `GetElements` component rather than having a separate component?

_the feature_ --> being able to create a pre-populated value list.

kovacsv commented:
> @tlorantfy, do you think it is better to have _"the feature"_ inside the `GetElements` component rather than having a separate component?
> 
> _the feature_ --> being able to create a pre-populated value list.

I'd prefer to have it in the "ElemsByType" component, too.

tlorantfy commented:
> @tlorantfy, do you think it is better to have _"the feature"_ inside the `GetElements` component rather than having a separate component?
> 
> _the feature_ --> being able to create a pre-populated value list.

Good idea. This valuelist component does not have standalone value.
Currently 2 components have element type input:
GetElemsByType and GetSubelements

I will do the change, so when we place a new GetElemsByType or GetSubelements component, a valulist like this will be automatically created and attached to the input.

kovacsv commented:
> > @tlorantfy, do you think it is better to have _"the feature"_ inside the `GetElements` component rather than having a separate component?
> 
> > 
> 
> > _the feature_ --> being able to create a pre-populated value list.
> 
> 
> 
> Good idea. This valuelist component does not have standalone value.
> 
> Currently 2 components have element type input:
> 
> GetElemsByType and GetSubelements
> 
> 
> 
> I will do the change, so when we place a new GetElemsByType or GetSubelements component, a valulist like this will be automatically created and attached to the input.

It is possible to list all the values on the context menu of an input. I think that's a more natural way of providing value list for an input. If that works I won't add another node.

tlorantfy commented:
@thekaushikls , @kovacsv , this is what I was able to do:
When we place a new GetElemsByType or GetSubelements component it automatically creates a pre-populated value list attached to the input.

https://github.com/user-attachments/assets/8022a900-5ba6-4c74-806c-f9c6320d262c

kovacsv commented:
I'd prefer doing this without an extra node. Here is an example of adding enum values to the context menu: https://discourse.mcneel.com/t/writing-a-very-simple-value-selector-in-visualstudio/102234/2

I don't know how it works with a lot of values, but if it handles that well i'd prefer to go with the context menu solution.

kovacsv commented:
> > I'd prefer doing this without an extra node. Here is an example of adding enum values to the context menu: https://discourse.mcneel.com/t/writing-a-very-simple-value-selector-in-visualstudio/102234/2
> 
> > 
> 
> > I don't know how it works with a lot of values, but if it handles that well i'd prefer to go with the context menu solution.
> 
> 
> 
> I have already checked that, but that works only for Integer parameters.
> 
> Element type is a string parameter.
> 
> We can change the input type of the component, but then we should change the Tapir Commands also.
> 
> 
> 
> For example GetDetailsOfElements command returns the type of the elements as string. Why dont attach that output to the input of GetElementsByType?

Oh, ok. Then there is no better solution than a separate node. One last question: is it possible to disconnect the node later?

tlorantfy commented:
> question: is it possible to disconnect the node later?

Yes, the created valuelist can be deleted/disconnected anytime.

kovacsv commented:
> > question: is it possible to disconnect the node later?
> 
> 
> 
> Yes, the created valuelist can be deleted/disconnected anytime.

And i guess you can create the selector node on its own, right?

kovacsv commented:
I've tried out and for me it appears like this after placement.
![image](https://github.com/user-attachments/assets/8a9efc76-a0d7-402b-8296-7af92759aaff)

Also, I can't find it among the nodes to place a new one if needed.

thekaushikls commented:
> > > I'd prefer doing this without an extra node. Here is an example of adding enum values to the context menu: https://discourse.mcneel.com/t/writing-a-very-simple-value-selector-in-visualstudio/102234/2
> > 
> > > 
> > 
> > > I don't know how it works with a lot of values, but if it handles that well i'd prefer to go with the context menu solution.
> > 
> > 
> > 
> > I have already checked that, but that works only for Integer parameters.
> > 
> > Element type is a string parameter.
> > 
> > We can change the input type of the component, but then we should change the Tapir Commands also.
> > 
> > 
> > 
> > For example GetDetailsOfElements command returns the type of the elements as string. Why dont attach that output to the input of GetElementsByType?
> 
> Oh, ok. Then there is no better solution than a separate node. Two questions: Is it possible to disconnect the node later? Is it possible to place the value selector node on its own (I don't see it among the tapir nodes in your latest video)?


I would (as a GH user) prefer the value list approach more. Mainly because it is convenient when I want to select more than one type simultaneously.

If such an option is not a priority, then the context menu is simpler to use.

JBENEITEZ commented:
> > > > I'd prefer doing this without an extra node. Here is an example of adding enum values to the context menu: https://discourse.mcneel.com/t/writing-a-very-simple-value-selector-in-visualstudio/102234/2
> > > 
> > > 
> > > > 
> > > 
> > > 
> > > > I don't know how it works with a lot of values, but if it handles that well i'd prefer to go with the context menu solution.
> > > 
> > > 
> > > I have already checked that, but that works only for Integer parameters.
> > > Element type is a string parameter.
> > > We can change the input type of the component, but then we should change the Tapir Commands also.
> > > For example GetDetailsOfElements command returns the type of the elements as string. Why dont attach that output to the input of GetElementsByType?
> > 
> > 
> > Oh, ok. Then there is no better solution than a separate node. Two questions: Is it possible to disconnect the node later? Is it possible to place the value selector node on its own (I don't see it among the tapir nodes in your latest video)?
> 
> I would (as a GH user) prefer the value list approach more. Mainly because it is convenient when I want to select more than one type simultaneously.
> 
> If such an option is not a priority, then the context menu is simpler to use.

I agree @thekaushikls, could you provide a sample code on how to autofill the values of the value list (like elefront does)?

kovacsv commented:
Thanks for the inputs, let's go with the value list. We just need to figure out tehse two things:
- The layout problem I posted earlier.
- The value list node doesn't appear among the nodes so I can't place a new one independently.

tlorantfy commented:
> * The layout problem I posted earlier.

Are you using Rhino7? I tested in Rhino8.

> * The value list node doesn't appear among the nodes so I can't place a new one independently.

I disallowed that intentionally (not assigned Guid to the valuelist and made constructor private).
It has no standalone value. If you think we should allow that, then I simply allow that.

kovacsv commented:
> > * The layout problem I posted earlier.
> 
> 
> 
> Are you using Rhino7? I tested in Rhino8.
> 
> 
> 
> > * The value list node doesn't appear among the nodes so I can't place a new one independently.
> 
> 
> 
> I disallowed that intentionally (not assigned Guid to the valuelist and made constructor private).
> 
> It has no standalone value. If you think we should allow that, then I simply allow that.

I'm using R8. I have a HighDPI monitor, maybe that's the problem. I would allow the standalone creation so if someone deletes it by accident it can be recreated.

kovacsv commented:
Looks like the width calculation is wrong in Grasshopper. It is based on the longest value, but the built-in value list adds the same result:
![image](https://github.com/user-attachments/assets/156c2326-57c7-4788-ab3c-8f634bb1e735)

Maybe if we can remove the "ElemType" part it will be better:
![image](https://github.com/user-attachments/assets/588a08bd-238f-4c11-ae88-770c99dda8df)

Still don't understand why the layouting doesn't work.

tlorantfy commented:
@kovacsv , I allowed the standalone creation of the valuelist, removed the "ElemType" nickname and adjusted the positioning using the scale of the display.
Could you please review?

kovacsv reviewed (approved):
Great improvement, thanks!
